SubGame_Network-1 Milestone Delivery

xtony77 opened the pull request:
# Milestone Delivery Checklist

- [x] The [milestone-delivery-template.md](https://github.com/w3f/Grant-Milestone-Delivery/blob/master/deliveries/milestone-delivery-template.md) has been copied and updated.
- [x] The [invoice form :pencil:](https://forms.gle/8Wx7nxtq8fKrsuEz8) has been filled out for this milestone. 
- [x] This pull request is being made by the same account as the accepted application. 
- [x] In case of acceptance, the payment will be transferred to the initial BTC/ETH payment address. 
- [x] The delivery is according to the [Guidelines for Milestone Deliverables](https://github.com/w3f/General-Grants-Program/blob/master/grants/milestone-deliverables-guidelines.md).

TomaszWaszczyk commented on deliveries/SubGame_Network-1.md:
```diff
@@ -0,0 +1,32 @@
+
+* **PR Link:** 
+    - https://github.com/w3f/Open-Grants-Program/pull/133
+    - https://github.com/w3f/Open-Grants-Program/pull/134
```
Look like the link does not work:

![image](https://user-images.githubusercontent.com/1605705/116795568-49094680-aad6-11eb-8c45-f43c90afe5a0.png)

xtony77 replied:
Fixed. Provide recent PR.

Noc2 commented:
Just want to specifically mention the evaluation PR here https://github.com/w3f/Grant-Milestone-Delivery/pull/163 You might want to take a look at it.

TomaszWaszczyk commented:
I think that meaning of evaluation is clear feedback on every stage as much as possible which is crucial in any form of communication. [Here](https://github.com/w3f/Grant-Milestone-Delivery/blob/master/evaluations/subgame_milestone-1_tomaszwaszczyk.md) I have written my comments and doubts about the project and white paper. 

Thanks @xtony77 for improvement, after fetching updated version I have successfully run the front-end and Go's back-end, good job!

![image](https://user-images.githubusercontent.com/1605705/117033303-881bdf80-ad02-11eb-84b6-297056f4a381.png)

In the white paper is written:

> After smart contract is processed by AI cloud computing, it can be perfectly compatible with other public chains to achieve high chain-to-chain collaboration.

It needs further deeper clarification what does it really mean, for purpose of clear communication would be good to see even draft of an architecture of promised solution, how are you going to implement it, it allows us to evaluate planned work.

I was looking at `Dockerfile` and suggest to use newer version of NodeJS (currently is `12.16.2`) even for security reasons but also best practice is to use latest stable version for new project, and of course SubGame is developed during last days and weeks ;-)

xtony77 commented:
A brief explanation of evaluation [#163 ](https://github.com/w3f/Grant-Milestone-Delivery/pull/163)  

1. **Documentation**
   * Name of the file shall be English-like, look at General Notes section : 
   The filename has been changed to English, thanks for your reminder 
   * providing unlimited possibilities for distributed mobile applications : 
   We will revise the white paper to make it more precise and specific.
   * SubGame is a public chain development team based on the Polkadot parachain :
   We will revise the white paper to make it more precise and specific.
   * It combines public cloud and private cloud hybrid storage technology & blockchain storage space : 
   Taking Guess Hash as an example, we will put game records and transaction data on the blockchain (public cloud), and also store the same data in the database (private cloud). In order to speed up the access, we will directly download Access in the private cloud, but users can also query the records on the public cloud at the same time;  IPFS is a good way to effectively solve the problem of data storage and has good compatibility with the blockchain. Our team will consider it in depth.
   * Polkadot can be said to be a creation public chain that has reached a new level of technical strength. It was developed by the Web3 Foundation :　
   We will revise the white paper to make it more precise and specific.
   * After the smart contract developed on the SubGame chain is processed by AI cloud computing, it can be perfectly compatible with other public chains to achieve high chain-to-chain collaboration. Could you describe more about "combining" smart contracts and AI, I know those word but what does it mean in practical sense? and other questions about cross-chain.:　
   The AI cloud computing in our document refers to the operation logic of the smart contract. As mentioned earlier, we actually want to describe the cross-chain ecological vision of joining Polkadot, so we originally referred to "Polkadot: vision for a heterogeneous multi-chain framework" The solution described in ", our current solution is to make a centralized system and interact with smart contracts of different chains, but also pay attention to the solutions of other teams in the entire Polkadot ecosystem.
   * In the whitepaper there are pictures of mobile phones, are you going to deliver mobile version, if so how are you going to integrate wallet on mobile? :　
   The white paper is only an example, but of course mobile phone applications may be constructed, only the presentation of the user interface is different.
   * Explanation of RPC in the white paper is described as "RPC refers to remote procedure call, that is to say, two servers A and B, one application is deployed on server A, and you want to call the functions/methods provided by the application on server B.", this is not exact explanation of RPC. : 
   Thank you for your advice, we will fix it.
   * I just want to ask about staging environment, there is any? If not, are you going to run the node with working frontend for public usage? : 
   We will prepare a test and a formal environment, which is currently in preparation.
   * I suggest also to improve bibliography in the whitepaper, some tips on that. : 
   Thank you for your reminder, we will make adjustments when the white paper is updated.
2. **Testing Guide**
   Taiwanese [comments](http://www.google.com/) and lack of Rustdoc : 
   Rustdocs comments have been added to the key parts and added english comments.
3. **SubGame Node**
   Taiwanese comments and lack of Rustdoc [link](https://github.com/SubGame-Network/subgame-network/commit/cf903db5256d0b75eed9fb1868d0a76e5bfa8123) :　
   Rustdocs comments have been added to the key parts and added english comments.
4. **pallet-chips**
   Hard coding weight value is not the best possible strategy vide [default_weight.rs](http://default_weight.rs/), to implement it better you can look at Substrate code where is used Frame System pallet, but leave it now and improve it in at a later stage. Let's add Rustdoc and remove Taiwanese comments : 
   Hard coding weight value improve it in at a later stage.
5. **pallet-gametemplates**
   Let start using Rustdoc, please improve readability of a source code like unified indentations : 
   The code has been a little optimized.
6. **pallet-gamecenter**
    Like above, could you explain why do you use ```extern crate alloc```? : 
    At that time, was used for test, but it is no longer useful. I have deleted the unused crates.thanks!
7. **pallet-gametemplates-guess-hash**
    Mixed type of comments - English and Taiwanese : 
    done.
8. **Front End**
    I am not able to run the [frontend part](https://github.com/SubGame-Network/GameCenter/issues/1), below I include more details. : 
    The frontend part can not run alone, you need the backend function to be a complete application, so it is recommended that you use docker to run.
9. **UI mock-ups**
    Which wireframes are legit one? : 
    The [wireframes](https://www.figma.com/file/hbwDsOVkP5tJqCnl7v0Smr/Subgame-center) is finally legal. Because the initial design screen has dice, we don’t want to cause misunderstandings, and we want to provide a better user interface, and closer to our pallet architecture, so we doc: [update on 4/16](https://github.com/w3f/Open-Grants-Program/pull/374/commits/5720343aa1ca3ae8c74dbd5e7c10fa139a42993a).

TomaszWaszczyk commented:
Thanks for the comments, I think that after rewriting whitepaper in more technical/more precise language it could be accepted for further implementation. Regarding RPC I have found some document, please take a look at: RPC on [ResearchGate](https://www.researchgate.net/publication/1959189_Enhanced_Prolog_Remote_Predicate_Call_Protocol) the idea is to communication like "CLIENT <==> SERVER", the document is extended but the picture is the most crucial part.

By filled with curiosity as I remember there were issues with calling Parity's RPC due to high load and the result was no possibility to look at accounts in polkadot.js, so the conclusion is that in the communication there is no need to have two servers RPC. In the included document you can see that there is one server and many clients, one writen in Prolog (as an example) and second one client written in Java.

Regarding front end part, mea culpa, sorry, I was thinking that it is possible to run (although without data) only front end part, mea culpa, I understand that. Well, according cool wire frames you have a lot of work to be done so I am not going to have more issues.

@Noc2  I would appreciate review of the review, comments and any feedback, I am open for any input.

mmagician commented:
@TomaszWaszczyk @xtony77 Thanks both of you for being very responsive.
A few quick comments from my side:
- the formatting of the code is still off, e.g. [here](https://github.com/SubGame-Network/subgame-network/blob/master/pallets/gametemplates/src/lib.rs#L55). Seeing that you don't have any preference for a style guideline, I'd suggest using [rustfmt](https://github.com/rust-lang/rustfmt)
- your repo contains an `ev` directory, where db & secrets files can be found. Unless they are necessary for some testing environment, I'd recommend removing it
- running `cargo clippy` produces multiple warnings. You should probably take a look at these as well and optimise your code
- you can remove the `template` pallet
- some pallets are missing unit tests

xtony77 commented:
> @TomaszWaszczyk @xtony77 Thanks both of you for being very responsive.
> A few quick comments from my side:
> 
> * the formatting of the code is still off, e.g. [here](https://github.com/SubGame-Network/subgame-network/blob/master/pallets/gametemplates/src/lib.rs#L55). Seeing that you don't have any preference for a style guideline, I'd suggest using [rustfmt](https://github.com/rust-lang/rustfmt)
> * your repo contains an `ev` directory, where db & secrets files can be found. Unless they are necessary for some testing environment, I'd recommend removing it
> * running `cargo clippy` produces multiple warnings. You should probably take a look at these as well and optimise your code
> * you can remove the `template` pallet
> * some pallets are missing unit tests

@mmagician Thanks for your comments, we have corrected these issues.

TomaszWaszczyk commented:
@xtony77 FYI: In order to work in more time efficient way with linters etc. etc. I have found how to implement hooks in Rustic way and automatize commands like `cargo clippy`, take a look: https://github.com/rhysd/cargo-husky

You can configure hook e.g. `pre-commit` hook and configure that without successful execution of `cargo clippy` there will be no way to commit new lines of source code. In that way it helps to keep source code more cleaner.

I am just sharing my experience from TypeScript na JS world, hope it helps ;-)

Noc2 commented:
Hi @xtony77 I very briefly also looked at the delivery and have two additional comments: 

- The testing guide for the pallets seems to be missing some information. Ideally even people unfamiliar with substrate should be able to install, compile and run the project. Could you add this to the md file? It generally also helps to add things like the rust version, which you are using, etc. 
- I’m also unable to run the frontend without docker. Is there a possibility to test it without docker, since everything should be open source for this delivery?

xtony77 commented:
> Hi @xtony77 I very briefly also looked at the delivery and have two additional comments:
> 
> * The testing guide for the pallets seems to be missing some information. Ideally even people unfamiliar with substrate should be able to install, compile and run the project. Could you add this to the md file? It generally also helps to add things like the rust version, which you are using, etc.
> * I’m also unable to run the frontend without docker. Is there a possibility to test it without docker, since everything should be open source for this delivery?

@Noc2 

- Thank you for the suggestion, we have added the md file with the information about the pallets
    - [rust-setup.md](https://github.com/SubGame-Network/subgame-network/blob/master/doc/rust-setup.md)
    - [pallets information example](https://github.com/SubGame-Network/subgame-network/blob/master/pallets/chips/README.md), you can find it in the other pallets folder
- Our architecture is designed to be complemented by a backend API. If you want to run the frontend alone, you can follow these steps, but some data will be missing:
    - copy `src/config/config.prod.json` to `src/config/config.json`
    - try run `yarn start`

TomaszWaszczyk commented:
@xtony77 Thanks for the updates, I think that (even for someone who wants to solve some issue voluntary for SubGame) to add `init.sh` script [here](https://github.com/SubGame-Network/subgame-network/tree/master/scripts)  to make easier building the development environment (for someone who wants to build from scratch) like it is implemented originally in Substrate framework [here](https://github.com/substrate-developer-hub/substrate-node-template/blob/master/scripts/init.sh).

After changing `src/config/config.prod.json` to `src/config/config.json` frontend works:

![image](https://user-images.githubusercontent.com/1605705/118644910-937d0980-b7de-11eb-84e0-8a2ebb0d07b8.png)

I think that in order to attract new developers building process should be as easy as possible, I exactly remember when I was building Substrate first time and on every step there was some 'hack', easy for those who knew it, but hard and really, really time consuming for newbie which has daily job and a lot of tasks. 

I suggest to write down hints how to run frontend explicitly [here](https://github.com/SubGame-Network/GameCenter/tree/master/frontend) in order to help those developers who wants to run it on own machine, build it from scratch and maybe solve some issue and/or experiment with the source code on own machine without Docker, and according to some best practices it can be implement also by using two Yarn scripts:
1. 'yarn start' - running the app locally
2. `yarn docker` - if needed for docker, just to keep behavior predictability of the code (most frontend devs assume that `yarn start` will run an app) 

or just update [file](https://github.com/SubGame-Network/GameCenter/tree/master/frontend#readme) with your instructions (copy `src/config/config.prod.json` to `src/config/config.json`)

All I can say is that we should be aware that developers do not have much time and building/running process without Docker should be as easy as possible and it eases development of a new features and Docker is like a box and me personally as a software engineer and does not attract to come and solve some issues even for free from simply reason - time consuming running development environment.

That is all, I am not going to be a blocker, do not know what you think @Noc2 but that is all what can I say.

xtony77 commented:
@TomaszWaszczyk we updated frontend instructions [here](https://github.com/SubGame-Network/GameCenter/tree/master/frontend#readme)

TomaszWaszczyk commented:
@xtony77 I have created simple [PR](https://github.com/SubGame-Network/subgame-network/pull/1) it just configures Rust's environment for those who want to build the node as a first try in Rust's world.

xtony77 commented:
@TomaszWaszczyk Thinks, [PR](https://github.com/SubGame-Network/subgame-network/pull/1) has been merged.

Noc2 commented:
@xtony77 Thanks for the updates. Given @TomaszWaszczyk great work (see [here](https://github.com/w3f/Grant-Milestone-Delivery/blob/master/evaluations/subgame_milestone-1_tomaszwaszczyk.md)). I’m happy to confirm that the milestone is approved and I will forward your invoice internally.

xtony77 commented:
@Noc2 Thank you all!

@TomaszWaszczyk About [evaluation](https://github.com/w3f/Grant-Milestone-Delivery/blob/master/evaluations/subgame_milestone-1_tomaszwaszczyk.md) I want to ask a few questions to do better:

- we use `cargo doc` to generate docs [here](https://github.com/SubGame-Network/subgame-network/blob/master/pallets/chips/README.md#documentation), but you still said lack of Rustdoc, what should we do to make it right?
- `cargo clippy` will warn some substrate errors, like the suggestion to remove `decl_event!()` without events, so we are not going to put in hooks, has anyone encountered similar situation? Any suggestions?

TomaszWaszczyk commented:
@xtony77 [Here](https://doc.rust-lang.org/stable/rust-by-example/meta/doc.html) you can find examples of Rustdoc, by writing Rustdoc I mean `///` syntax and short description of functions with the syntax.

Regarding `cargo clippy`, first of all I want to emphasize that my assumption was to show it, I do not believe in perfect code and I do believe that it is better to finish than implement something in ideal way, definitely better done than perfect.

I was thinking about solving it and got to the conclusion (not sure that the best one) that the best solution for that will be implementation of custom rules for `cargo clippy` - not all warnings make sense in context of Substrate. Specifically referring to `decl_event!()` I think addition of `#[allow(clippy::unused_unit)]` should help, but as I said I think that the topic of customized `cargo clippy` requires more work and to be honest it is not a crucial, although I think that work in order to get some best practices would be helpful for future builders.

In order to remember it I have created [question](https://stackoverflow.com/questions/67625287/how-do-i-hide-the-warning-from-clippy-about-unneeded-unit-expression-when-usin), in my opinion over time delivering real value those rules will just appear and the art will be to collect them in one place (probably in the form of rules for `cargo-clippy`) and then share along the wider audience/builders. Just to collect the knowledge and spread it in efficient way.

xtony77 commented:
@TomaszWaszczyk Thank you for your reply, it was very helpful and we will continue to optimize SubGame!

RouvenP commented:
hi @xtony77 we transferred the payment last Friday.

xtony77 commented:
@RouvenP we confirm received, thank you.

TomaszWaszczyk commented:
@Noc2 @mmagician Thanks for the chance to get to know the process, I have learnt something new, it is high time for me to work more around Substrate and all things around, I tried the topic rather than lifetime journey than short incident.

Best,
Tomek
;-(

#keepworking
